cheroot/test/test_ssl.py

```diff
@@ -775,31 +780,116 @@ def test_http_over_https_error(
+)
+def test_send_bad_request_socket_errors(
+    mocker,
+    error,
+    raising_expectation,
+):
+    """Test socket error handling when sending 400 Bad Request."""
+    # Mock the selector in ConnectionManager initialization
+    mocker.patch('cheroot.connections._ThreadsafeSelector', autospec=True)
+
+    mock_server = mocker.Mock(spec=HTTPServer)
+    mock_server.protocol = 'HTTP/1.1'
+    mock_server.socket = mocker.Mock(spec=socket.socket)
+    conn_manager = ConnectionManager(mock_server)
+
+    mock_raw_socket = mocker.Mock(spec=socket.socket)
+    mock_raw_socket.sendall.side_effect = error
+
+    with raising_expectation as exc_info:
+        conn_manager._send_bad_request_plain_http_error(
+            mock_raw_socket,
         )
 
-    underlying_error = ssl_err.value.args[0].args[-1]
-    err_text = str(underlying_error)
-    assert underlying_error.errno == expected_error_code, (
-        'The underlying error is {underlying_error!r}'.format(**locals())
+    # If we expect an error, check it's the correct one
+    should_propagate = not isinstance(
+        raising_expectation,
+        contextlib.nullcontext,
     )
-    assert expected_error_text in err_text
+    if should_propagate:
+        assert exc_info.value is error
+
+    mock_raw_socket.sendall.assert_called_once()
+    mock_raw_socket.close.assert_called_once()
+
+
+@pytest.mark.parametrize('adapter_type', ('builtin', 'pyopenssl'))
+# pylint: disable-next=too-many-positional-arguments
+def test_http_over_https_ssl_handshake(
+    mocker,
+    tls_http_server,
+    adapter_type,
+    tls_certificate,
+    tls_certificate_chain_pem_path,
+    tls_certificate_private_key_pem_path,
+):
+    """
+    Test NoSSLError raised when SSL handshake catches HTTP.
+
+    Normally the early probe ``_ensure_peer_speaks_https()``
+    will detect a client attempting to speak HTTP on a TLS
+    port but if this times out or fails for some reason, SSL
+    should raise an error at the time of the handshake. Here
+    we test the error is caught and triggers the emission of
+    a ``400 Bad Request``.
+    """
+    interface, _host, port = _get_conn_data(ANY_INTERFACE_IPV4)
+
+    tls_adapter_cls = get_ssl_adapter_class(name=adapter_type)
+    tls_adapter = tls_adapter_cls(
+        tls_certificate_chain_pem_path,
+        tls_certificate_private_key_pem_path,
+    )
+
+    tls_certificate.configure_cert(tls_adapter.context)
+
+    # Mock the early probe to not detect the HTTP request
+    mocker.patch('cheroot.ssl._ensure_peer_speaks_https')
+
+    tlshttpserver = tls_http_server((interface, port), tls_adapter)
+
+    interface, _host, port = _get_conn_data(tlshttpserver.bind_addr)
+
+    # Send plain HTTP
+    with socket.create_connection((interface, port)) as sock:
+        sock.sendall(b'GET / HTTP/1.1\r\nHost: localhost\r\n\r\n')
+
+        if adapter_type == 'pyopenssl':
+            # pyopenssl can send the 400 response
+            response = sock.recv(4096)
```

Comment:
Should we just read what we're checking for?
```suggestion
            response = sock.recv(15)
```

Reply:
@julianz- any opinion here?